src/core/DataFile.cpp

```diff
@@ -328,6 +405,164 @@ bool DataFile::writeFile( const QString& filename )
+	std::list<QString> namesList;
+
+	ResourcesMap::const_iterator it = ELEMENTS_WITH_RESOURCES.begin();
+
+	// Copy resources and manipulate the DataFile to have local paths to them
+	while (it != ELEMENTS_WITH_RESOURCES.end())
+	{
+		QDomNodeList list = elementsByTagName(it->first);
+
+		// Go through all elements with the tagname from our map
+		for (int i = 0; !list.item(i).isNull(); ++i)
+		{
+			QDomElement el = list.item(i).toElement();
+
+			std::vector<QString>::const_iterator res = it->second.begin();
+
+			// Search for attributes that point to resources
+			while (res != it->second.end())
+			{
+				// If the element has that attribute
+				if (el.hasAttribute(*res))
+				{
+					// Get absolute path to resource
+					QString resPath = PathUtil::toAbsolute(el.attribute(*res));
+					// If we are running without the project loaded (from CLI), "local:" base
+					// prefixes aren't converted, so we need to convert it ourselves
+					if (PathUtil::baseLookup(resPath) == PathUtil::Base::LocalDir)
+					{
+						resPath = QFileInfo(m_fileName).path() + "/" + resPath.remove(0,
+							PathUtil::basePrefix(PathUtil::Base::LocalDir).length());
+					}
+
+					// Check if we need to add a counter to the filename
+					QString finalFileName = QFileInfo(resPath).fileName();
+					QString extension = resPath.section('.', -1);
+					int repeatedNames = 0;
+					for (QString name : namesList)
+					{
+						if (finalFileName == name)
+						{
+							++repeatedNames;
+						}
+					}
+					// Add the name to the list before modifying it
+					namesList.push_back(finalFileName);
+					if (repeatedNames)
+					{
+						// Remove the extension, add the counter and add the
+						// extension again to get the final file name
+						finalFileName.truncate(finalFileName.lastIndexOf('.'));
+						finalFileName = finalFileName + "-" + QString::number(repeatedNames) + "." + extension;
+					}
+
+					// Final path is our resources dir + the new file name
+					QString finalPath = resourcesDir + "/" + finalFileName;
+
+					// Copy resource file to the resources folder
+					if(!QFile::copy(resPath, finalPath))
+					{
+						qWarning("ERROR: Failed to copy resource");
+						return false;
+					}
+
+					// Update attribute path to point to the bundle file
+					QString newAtt = PathUtil::basePrefix(PathUtil::Base::LocalDir) + "resources/" + finalFileName;
+					el.setAttribute(*res, newAtt);
+				}
+				++res;
+			}
+		}
+		++it;
+	}
+
+	return true;
+}
+
+
+
+
+/* @brief This recursive method will go through all XML nodes of the DataFile
+ *        and check whether any of them have local paths. If they are not on
+ *        our list of elements that can have local paths we return true,
+ *        indicating that we potentially have plugins with local paths that
+ *        would be a security issue. The Song class can then abort loading
+ *        this project.
+ * @param QDomElement parent The parent node being interated. When called
```

Comment:
Also, is there a reason why you specified the types after `@param` in this comment? I haven't seen that in doxygen yet.

Reply:
I've seen it before on `AutomationPattern.cpp`, so I thought it was common to have the types in the `@param` comment. I wrote as I've seen them there: "type name description".

Reply:
If you type `make doc`, you could see how it works. However, it currently does only display the text from the header, which is wrong. Probably, you either need to move it into the header or into the cpp file.

Reply:
`make doc` didn't work here for some reason, but anyways, you mean that because there's a Doxygen comment on the header and on the cpp file at the same time, only the header comment is being used in the docs? So I should just remove it from the header?

Reply:
I think that's the reason, yes. Remove it from header or from cpp file. But both at the same time will not work. You should really try to get `make doc` running or report otherwise.

Reply:
I didn't have `doxygen` installed, my bad :man_facepalming: . I removed the doxygen comment from the header and left only the source file one. It appears to have build into the document properly. It also seems like doxygen indeed identifies the first word after `@param` as the variable type, as it's colored differently.